keps/sig-apps/4443-configurable-pod-failure-policy-reasons/README.md

```diff
@@ -0,0 +1,971 @@
+we are validating against malformed/invalid inputs.
+
+
+## Design Details
+
+#### Defaulting
+If unset, the `Name` field will default to the index of the pod failure policy rule in the `Rules` slice.
+
+#### Validation
+- We will validate the name be used as a valid Reason ([non-empty, CamelCase string](https://github.com/kubernetes/kubernetes/blob/dd301d0f23a63acc2501a13049c74b38d7ebc04d/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1555)). We will also validate the user-defined Reason is less than 128 characters.
+- We will also validate the pod failure policy rule does not conflict with any [K8s internal reasons used by the Job controller](https://github.com/kubernetes/kubernetes/blob/862ff187baad9373d59d19e5d736dcda1e25e90d/staging/src/k8s.io/api/batch/v1/types.go#L542-L553). 
+
+#### Business logic
+When a `PodFailurePolicyRule` matches a pod failure and the `Action` is `FailJob`, the Job controller will
+set the JobFailed condition reason deterministically in the format `PodFailurePolicy_{ruleName}`. This 
+suffix will be added to the condition [condition](https://github.com/kubernetes/kubernetes/blob/6a4e93e776a35d14a61244185c848c3b5832621c/pkg/controller/job/job_controller.go#L816) here.
+
+Note: If the PodFailurePolicy feature gate is disabled, but the `PodFailurePolicyName` feature gate is enabled, there will be 
+no adverse effect and neither feature will actually be used, since the only place the proposed new field `Name` will be
+used is inside of a [code block](https://github.com/kubernetes/kubernetes/blob/6a4e93e776a35d14a61244185c848c3b5832621c/pkg/controller/job/job_controller.go#L816) protected by the PodFailurePolicy feature gate.
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `k8s.io/kubernetes/pkg/controller/job`: `02/05/2024` - `91.5%`
```

Comment:
What about the batch api validation packages?

Reply:
This still holds.

Reply:
Added batch api and batch validation packages.